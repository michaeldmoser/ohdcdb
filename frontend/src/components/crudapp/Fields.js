import { NavLink } from 'react-router-dom';
import PropTypes from 'utils/PropTypes';
import _ from 'lodash';

/**
 * Display a field on the detailview
 *
 * @param {string} props.label - The user visible name of the field
 * @param {*} props.value - The value to be rendered for the field
 * @returns
 */
export const Field = ({ label, value, ...props }) => {
    return (
        <>
            <dt className='col-sm-3' aria-label={label}>
                {label}
            </dt>
            <dd className='col-sm-9' aria-label={value}>
                {value}
            </dd>
        </>
    );
};
Field.propTypes = {
    label: PropTypes.string.isRequired,
    value: PropTypes.any,
};

/**
 * Renders a container for Fields on a detail view
 *
 * @param {array} props.children - An array of Field components
 */
export const FieldContainer = ({ children, ...props }) => {
    return <dl className='row'>{children}</dl>;
};

FieldContainer.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.isComponent(Field)),
        PropTypes.isComponent(Field),
    ]),
};

/**
 * Renders as a basic field but lists items as links to other records in the database.
 *
 * @param {string|function} [props.idField=id] - Can be a string that is the key lookup or a function that returns the value to use as a unique key.
 * @param {string|function} [props.href] - Can be a string that is the key lookup of the unique identifier to use for each related item or a function that returns the value for the href property of the link.
 * @param {string} label - The user visible name of the field
 * @param {Object[]} value - An array of objects to be rendered in a list
 * @param {function} children - Function to render the value of each related item.
 *
 * @example
 * // <RelatedField label='Owners' value=[{id: 1, first_name: 'John', last_name: 'Doe'}, {id: 2, first_name: 'Jane', last_name: 'Doe'}] idField='id'>
 * // ((item) => <>{item.first_name} {item.last_name}</>)
 * // </RelatedField>
 *
 * @example
 * // <RelatedField label='Owners' value=[{id: 1, first_name: 'John'}...] idField=((item) => item.id)>
 * // ((item) => <>{item.first_name} {item.last_name}</>)
 * // </RelatedField>
 *
 * @example
 * // <RelatedField label='Owners' value=[{id: 1, first_name: 'John'}...] href=((item) => `/items/${item.id})`>
 * // ((item) => <>{item.first_name} {item.last_name}</>)
 * // </RelatedField>
 *
 * @example
 * // <RelatedField label='Owners' value=[{id: 1, first_name: 'John'}...] href='id'>
 * // ((item) => <>{item.first_name} {item.last_name}</>)
 * // </RelatedField>
 */
export const RelatedField = ({
    value,
    idField = 'id',
    href,
    children,
    ...props
}) => {
    return (
        <Field
            value={
                <ul className='list-unstyled'>
                    {value.map((related) => {
                        const key = _.isString(idField)
                            ? related[idField]
                            : _.isFunction(idField)
                            ? idField(related)
                            : related?.id;

                        const link = _.isFunction(href)
                            ? href(related)
                            : _.isEmpty(href)
                            ? key
                            : `${href}/${key}`;

                        return (
                            <li key={key}>
                                <NavLink to={`${link}`} className='lh-sm'>
                                    {children(related)}
                                </NavLink>
                            </li>
                        );
                    })}
                </ul>
            }
            {...props}
        />
    );
};

RelatedField.propTypes = {
    href: PropTypes.oneOfType([
        PropTypes.func.isRequired,
        PropTypes.string.isRequired,
    ]),
    idField: PropTypes.oneOfType([
        PropTypes.func.isRequired,
        PropTypes.string.isRequired,
    ]),
    label: PropTypes.string.isRequired,
    value: PropTypes.array.isRequired,
    children: PropTypes.func.isRequired,
};
