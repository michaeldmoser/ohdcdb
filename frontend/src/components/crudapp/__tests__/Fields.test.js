import React from 'react';

import { render } from 'testing/library';

import { RelatedField, FieldContainer, Field } from '../Fields';

describe('RelatedField', () => {
    const relations = [
        {
            id: 1,
            id2: 3,
            id3: 5,
            title: 'Title 1',
            description: 'Description 1',
        },
        {
            id: 2,
            id2: 4,
            id3: 6,
            title: 'Title 2',
            description: 'Description 2',
        },
    ];

    const child = ({ title, description }) => (
        <>
            {title} {description}
        </>
    );

    it('should render using default id value', () => {
        const rendered = render(
            <RelatedField value={relations} label='Related'>
                {child}
            </RelatedField>
        );

        expect(rendered).toMatchSnapshot();
    });

    it('should render using string id', () => {
        const rendered = render(
            <RelatedField value={relations} label='Related' idField='id2'>
                {child}
            </RelatedField>
        );

        expect(rendered).toMatchSnapshot();
    });

    it('should render with idField as a function', () => {
        const rendered = render(
            <RelatedField
                value={relations}
                label='Related'
                idField={({ id3 }) => id3}
            >
                {child}
            </RelatedField>
        );

        expect(rendered).toMatchSnapshot();
    });

    it('should render with href as a string', () => {
        const rendered = render(
            <RelatedField value={relations} label='Related' href='/items'>
                {child}
            </RelatedField>
        );

        expect(rendered).toMatchSnapshot();
    });

    it('should render with href as a function', () => {
        const rendered = render(
            <RelatedField
                value={relations}
                label='Related'
                href={({ id2 }) => `/items/${id2}`}
            >
                {child}
            </RelatedField>
        );

        expect(rendered).toMatchSnapshot();
    });
});

describe('Fields and field', () => {
    it('should render a basic field label and value', () => {
        const rendered = render(
            <FieldContainer>
                <Field value={1} label='Field label' />
            </FieldContainer>
        );

        expect(rendered).toMatchSnapshot();
    });
});
