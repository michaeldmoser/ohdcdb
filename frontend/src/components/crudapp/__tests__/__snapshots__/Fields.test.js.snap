// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields and field should render a basic field label and value 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <dl
        class="row"
      >
        <dt
          aria-label="Field label"
          class="col-sm-3"
        >
          Field label
        </dt>
        <dd
          aria-label="1"
          class="col-sm-9"
        >
          1
        </dd>
      </dl>
    </div>
  </body>,
  "container": <div>
    <dl
      class="row"
    >
      <dt
        aria-label="Field label"
        class="col-sm-3"
      >
        Field label
      </dt>
      <dd
        aria-label="1"
        class="col-sm-9"
      >
        1
      </dd>
    </dl>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RelatedField should render using default id value 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <dt
        aria-label="Related"
        class="col-sm-3"
      >
        Related
      </dt>
      <dd
        aria-label="[object Object]"
        class="col-sm-9"
      >
        <ul
          class="list-unstyled"
        >
          <li>
            <a
              class="lh-sm"
              href="/1"
            >
              Title 1
               
              Description 1
            </a>
          </li>
          <li>
            <a
              class="lh-sm"
              href="/2"
            >
              Title 2
               
              Description 2
            </a>
          </li>
        </ul>
      </dd>
    </div>
  </body>,
  "container": <div>
    <dt
      aria-label="Related"
      class="col-sm-3"
    >
      Related
    </dt>
    <dd
      aria-label="[object Object]"
      class="col-sm-9"
    >
      <ul
        class="list-unstyled"
      >
        <li>
          <a
            class="lh-sm"
            href="/1"
          >
            Title 1
             
            Description 1
          </a>
        </li>
        <li>
          <a
            class="lh-sm"
            href="/2"
          >
            Title 2
             
            Description 2
          </a>
        </li>
      </ul>
    </dd>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RelatedField should render using string id 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <dt
        aria-label="Related"
        class="col-sm-3"
      >
        Related
      </dt>
      <dd
        aria-label="[object Object]"
        class="col-sm-9"
      >
        <ul
          class="list-unstyled"
        >
          <li>
            <a
              class="lh-sm"
              href="/3"
            >
              Title 1
               
              Description 1
            </a>
          </li>
          <li>
            <a
              class="lh-sm"
              href="/4"
            >
              Title 2
               
              Description 2
            </a>
          </li>
        </ul>
      </dd>
    </div>
  </body>,
  "container": <div>
    <dt
      aria-label="Related"
      class="col-sm-3"
    >
      Related
    </dt>
    <dd
      aria-label="[object Object]"
      class="col-sm-9"
    >
      <ul
        class="list-unstyled"
      >
        <li>
          <a
            class="lh-sm"
            href="/3"
          >
            Title 1
             
            Description 1
          </a>
        </li>
        <li>
          <a
            class="lh-sm"
            href="/4"
          >
            Title 2
             
            Description 2
          </a>
        </li>
      </ul>
    </dd>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RelatedField should render with href as a function 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <dt
        aria-label="Related"
        class="col-sm-3"
      >
        Related
      </dt>
      <dd
        aria-label="[object Object]"
        class="col-sm-9"
      >
        <ul
          class="list-unstyled"
        >
          <li>
            <a
              class="lh-sm"
              href="/items/3"
            >
              Title 1
               
              Description 1
            </a>
          </li>
          <li>
            <a
              class="lh-sm"
              href="/items/4"
            >
              Title 2
               
              Description 2
            </a>
          </li>
        </ul>
      </dd>
    </div>
  </body>,
  "container": <div>
    <dt
      aria-label="Related"
      class="col-sm-3"
    >
      Related
    </dt>
    <dd
      aria-label="[object Object]"
      class="col-sm-9"
    >
      <ul
        class="list-unstyled"
      >
        <li>
          <a
            class="lh-sm"
            href="/items/3"
          >
            Title 1
             
            Description 1
          </a>
        </li>
        <li>
          <a
            class="lh-sm"
            href="/items/4"
          >
            Title 2
             
            Description 2
          </a>
        </li>
      </ul>
    </dd>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RelatedField should render with href as a string 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <dt
        aria-label="Related"
        class="col-sm-3"
      >
        Related
      </dt>
      <dd
        aria-label="[object Object]"
        class="col-sm-9"
      >
        <ul
          class="list-unstyled"
        >
          <li>
            <a
              class="lh-sm"
              href="/items/1"
            >
              Title 1
               
              Description 1
            </a>
          </li>
          <li>
            <a
              class="lh-sm"
              href="/items/2"
            >
              Title 2
               
              Description 2
            </a>
          </li>
        </ul>
      </dd>
    </div>
  </body>,
  "container": <div>
    <dt
      aria-label="Related"
      class="col-sm-3"
    >
      Related
    </dt>
    <dd
      aria-label="[object Object]"
      class="col-sm-9"
    >
      <ul
        class="list-unstyled"
      >
        <li>
          <a
            class="lh-sm"
            href="/items/1"
          >
            Title 1
             
            Description 1
          </a>
        </li>
        <li>
          <a
            class="lh-sm"
            href="/items/2"
          >
            Title 2
             
            Description 2
          </a>
        </li>
      </ul>
    </dd>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RelatedField should render with idField as a function 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <dt
        aria-label="Related"
        class="col-sm-3"
      >
        Related
      </dt>
      <dd
        aria-label="[object Object]"
        class="col-sm-9"
      >
        <ul
          class="list-unstyled"
        >
          <li>
            <a
              class="lh-sm"
              href="/5"
            >
              Title 1
               
              Description 1
            </a>
          </li>
          <li>
            <a
              class="lh-sm"
              href="/6"
            >
              Title 2
               
              Description 2
            </a>
          </li>
        </ul>
      </dd>
    </div>
  </body>,
  "container": <div>
    <dt
      aria-label="Related"
      class="col-sm-3"
    >
      Related
    </dt>
    <dd
      aria-label="[object Object]"
      class="col-sm-9"
    >
      <ul
        class="list-unstyled"
      >
        <li>
          <a
            class="lh-sm"
            href="/5"
          >
            Title 1
             
            Description 1
          </a>
        </li>
        <li>
          <a
            class="lh-sm"
            href="/6"
          >
            Title 2
             
            Description 2
          </a>
        </li>
      </ul>
    </dd>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
